File: tree/determine_the_level_of_each_node_in_the_given_tree.py
tree = dict()
tree[0]=[1,2]
tree[1]=[3,4]
tree[2]=[5,6]
tree[3]=[7,8]
tree[4]=[]
tree[5]=[]
tree[6]=[]
tree[7]=[]
tree[8]=[]
def determine_node_level_in_tree(tree,root):
    node_level=dict()
    queue = [root]
    node_level[root]=0
    while len(queue) != 0 :
        p = queue.pop(0)
        for n in tree[p] :
            queue.append(n)
            node_level[n]=node_level[p]+1
    return node_level
print(determine_node_level_in_tree(tree,0))

File: tree/determine_node_level_in_tree_in_go/determine_the_level_of_each_node_in_the_given_tree.go
package main

import "fmt"

func main() {
	test_tree := map[int][]int{
		0: {1, 2},
		1: {3, 4},
		2: {5, 6},
		3: {7, 8},
		4: {},
		5: {},
		6: {},
		7: {},
		8: {},
	}
	fmt.Println(determine_node_level_in_tree(test_tree, 0))
}

func determine_node_level_in_tree(tree map[int][]int, root int) map[int]int {
	node_level := map[int]int{root: 0}
	queue := []int{}
	queue = append(queue, root) //enqueue
	var p int
	for len(queue) != 0 {
		p = queue[0]
		queue = queue[1:] //dequeue
		for _, n := range tree[p] {
			queue = append(queue, n)
			// key : decide level when adding candidate node in queue
			node_level[n] = node_level[p] + 1
		}
	}
	return node_level
}
